src/babel-serializer/network/pcap.hh

```diff
@@ -2,65 +2,114 @@
+header header_of(cc::span<std::byte const> data, read_config const& cfg = {}, error_handler on_error = default_error_handler);
+
+/// usage:
+/// auto data = ...;
+/// for(auto p : packets_of(data))
+///     my_process_data(p.data());
+///
+/// IMPORTANT: data MUST survive the packet_range and its iterators!
+/// IMPORTANT: on_error MUST survive the packet_range and its iterators!
+///
+/// Usage with a custom error handler:
+/// auto data = ...;
+/// auto my_error_handler = [&](auto data, auto pos, auto message, auto severity){...};
+/// auto cfg = babel::pcap::read_config();
+/// for(auto p : packets_of(data, cfg, my_error_handler))
+///     my_process_data(p.data());
+detail::packet_range packets_of(cc::span<std::byte const> data, read_config const& cfg = {}, error_handler on_error = default_error_handler);
+
+namespace detail
 {
-    // NOTE: this is the global header of a pcap file
-    struct header
-    {
-        uint32_t magic_number;  ///< magic number
-        uint16_t version_major; ///< major version number
-        uint16_t version_minor; ///< minor version number
-        int32_t thiszone;       ///< GMT to local correction
-        uint32_t sigfigs;       ///< accuracy of timestamps
-        uint32_t snap_length;   ///< max length of captured packets, in octets
-        uint32_t network;       ///< data link type
-    } header;
-
-    struct packet
+struct packet_iterator
+{
+    struct packet_header
     {
-        uint32_t timestamp_sec;  ///< timestamp seconds
-        uint32_t timestamp_usec; ///< timestamp microseconds
-        uint32_t original_size;  ///< original length of packet (NOT the stored one)
-
-        uint32_t stored_size; ///< actual stored length of the packet (might have been truncated)
-        size_t stored_offset; ///< offset into data::bytes
+        uint32_t ts_sec;   /* timestamp seconds */
+        uint32_t ts_usec;  /* timestamp microseconds */
+        uint32_t incl_len; /* number of octets of packet saved in file */
+        uint32_t orig_len; /* actual length of packet */
     };
 
-    /// vector of all packets
-    cc::vector<packet> packets;
-
-    /// returns a view on the packet data
-    /// NOTE: does not work if cfg.view_only was true
-    cc::span<std::byte const> packet_data(packet const& p) { return cc::as_byte_span(bytes).subspan(p.stored_offset, p.stored_size); }
+    packet_iterator(cc::span<std::byte const> data, read_config const& cfg, error_handler on_error, bool swap_endianness);
+    packet operator*() { return _current_packet; }
+    packet_iterator& operator++();
+    packet_iterator operator++(int)
+    {
+        auto i = *this;
+        operator++();
+        return i;
+    }
+    bool operator!=(cc::sentinel) const { return _pos < _data.size(); }
 
-    /// raw bytes of the pcap file (indexed by packet data)
-    cc::array<std::byte> bytes;
+private:
+    packet read_packet();
+    cc::span<std::byte const> _data;
+    read_config const& _cfg; // currently unused
```

Comment:
this is a dangling reference. just store it by value

Reply:
Fixed